### PR TITLE
ledger-api-client + participant-integration-api: Increase the default maximum inbound error size, and truncate errors well before that.

### LINKED ISSUE #5738: Sandbox: Header size exceeded max allowed size
In one of our tests, I get the following error from Sandbox:
```
WARNING: Stream Error
io.netty.handler.codec.http2.Http2Exception$HeaderListSizeException: Header size exceeded max allowed size (8192)
        at io.netty.handler.codec.http2.Http2Exception.headerListSizeError(Http2Exception.java:189)
        at io.netty.handler.codec.http2.Http2CodecUtil.headerListSizeExceeded(Http2CodecUtil.java:232)
        at io.netty.handler.codec.http2.HpackEncoder.encodeHeadersEnforceMaxHeaderListSize(HpackEncoder.java:122)
        at io.netty.handler.codec.http2.HpackEncoder.encodeHeaders(HpackEncoder.java:106)
        at io.netty.handler.codec.http2.DefaultHttp2HeadersEncoder.encodeHeaders(DefaultHttp2HeadersEncoder.java:68)
        at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeadersInternal(DefaultHttp2FrameWriter.java:501)
        at io.netty.handler.codec.http2.DefaultHttp2FrameWriter.writeHeaders(DefaultHttp2FrameWriter.java:268)
        at io.netty.handler.codec.http2.Http2OutboundFrameLogger.writeHeaders(Http2OutboundFrameLogger.java:60)
        at io.netty.handler.codec.http2.DecoratingHttp2FrameWriter.writeHeaders(DecoratingHttp2FrameWriter.java:53)
        at io.grpc.netty.NettyServerHandler$WriteMonitoringFrameWriter.writeHeaders(NettyServerHandler.java:1016)
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder.writeHeaders(DefaultHttp2ConnectionEncoder.java:208)
        at io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder.writeHeaders(DefaultHttp2ConnectionEncoder.java:151)
        at io.grpc.netty.NettyServerHandler.sendResponseHeaders(NettyServerHandler.java:683)
        at io.grpc.netty.NettyServerHandler.write(NettyServerHandler.java:602)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:716)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:708)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:791)
        at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:701)
        at io.netty.channel.DefaultChannelPipeline.write(DefaultChannelPipeline.java:1026)
        at io.netty.channel.AbstractChannel.write(AbstractChannel.java:288)
        at io.grpc.netty.WriteQueue$AbstractQueuedCommand.run(WriteQueue.java:200)
        at io.grpc.netty.WriteQueue.flush(WriteQueue.java:115)
        at io.grpc.netty.WriteQueue.access$000(WriteQueue.java:34)
        at io.grpc.netty.WriteQueue$1.run(WriteQueue.java:46)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:416)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:515)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)

```
What am I doing wrong?

Please look at the following pull request (InventoryManagementIT is failing):
https://github.com/digital-asset/ex-inventory-management/pull/26
(https://212-215083617-gh.circle-artifacts.com/0/logs/integration-test-sandbox.log)

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
Tested with the _HelloService_.

Fixes #5738.

### Changelog

- **[Integration Kit]** Truncate GPRC error messages at 4 KB. This ensures that we won't trigger a protocol error when sending errors to the client.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/TruncatedStatusInterceptor.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  private def truncate(description: String): String =
+    Option(description)
+      .map(
+        s =>
+          if (s.length > maximumDescriptionLength)
+            s.substring(0, maximumDescriptionLength - 3) + "..."
+          else
+          s)
+      .orNull
```
Comment (stefanobaghino-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/TruncatedStatusInterceptor.scala:33]:
```suggestion
      if (s != null && s.length > maximumDescriptionLength)
        s.substring(0, maximumDescriptionLength - 3) + "..."
      else
        s
```
Seems simpler, feel free to drop if we have some linting check that prevents us from doing it.

File: ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala [hunk 2]
```diff
@@ -3,18 +3,32 @@
...
+    maxInboundMessageSize: Int = DefaultMaxInboundMessageSize,
+)
+
+object LedgerClientConfiguration {
+
+  // This needs to be large because we often include parts of the request in GRPC errors, which are
+  // returned in the response headers. A large request could lead to the client choking on the
+  // response headers if the limit is too low.
+  // The default limit set by Netty is 8 KB, which often leads to errors.
+  val DefaultMaxInboundMessageSize: Int = 1024 * 1024 // 1 MB
```
Comment (stefanobaghino-da) [inline ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala:32]: Seems a bit high, do we really want to allow dumps so large? I think that the built-in limit is fine, as the dump is intended mostly for debugging purposes when developing small things. For larger dumps the process of inspecting the textual dump likely makes little sense.

Comment (SamirTalwar-DA) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala:32]: DO you think we should leave it to 8 KB then, and just truncate?

Comment (stefanobaghino-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala:32]: I think 8KB is fine. At that point, the dump is already basically unreadable and there should exist some other way of dealing with it.

Comment (gerolf-da) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala:32]: yeah, I think the error message should describe the error well. We can log the full transaction in DEBUG mode on the server side.
Depending on where the response header limit error is thrown, maybe we can install a generic interceptor that does the truncation at a safe length (a few hundred chars should be enough). This is to ensure that the user always gets to see the real error.

Comment (SamirTalwar-DA) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala:32]: That's exactly what the [`TruncatedStatusInterceptor`](https://github.com/digital-asset/daml/pull/6807/files#diff-969b0fe6e367db5e9d13cc1601f181d8) does. 🙂

Comment (SamirTalwar-DA) [reply ledger/ledger-api-client/src/main/scala/com/digitalasset/ledger/client/configuration/LedgerClientConfiguration.scala]: Reduced back to 8 KB (the default), and error messages are now truncated to 4 KB.

**Pull request merged**